### PR TITLE
Improve BigDecimal mongoize behavior

### PR DESCRIPTION
**Pull request by johnnyshields**
Recently I attempted to migrate a number of `Float` fields to `BigDecimal`, and I found that the behavior is inconsistent between the two. This PR rectifies that situation. Note that `BigDecimal` is stored in MongoDB as a string, as there is no native BSON type for it.

- `demongoize` should always return a BigDecimal or else nil, never an empty string (this mirrors how Floats and Integers behave).
- `mongoize` should always return a String which can be demongoized to a BigDecimal, otherwise nil
- Support storing BigDecimal values `"Infinity"` and `"-Infinity"` in the database.
- The strings `"Infinity"` and `"-Infinity"` now evaluate as numeric? true.
- `BigDecimal`s now evaluate as numeric? true.
- Additional tests for BigDecimal covering all edge cases.

### Demongoize

| Value to Demongoize | Current            | New | Behavior Changed? |
| ---         | ---                | --- | --- |
| "0"         | BigDecimal 0.0     | BigDecimal 0.0 |   |
| "-0"        | BigDecimal -0.0    | BigDecimal -0.0 |   |
| "1"         | BigDecimal 1       | BigDecimal 1 |   |
| "1.1"       | BigDecimal 1.1     | BigDecimal 1.1 |   |
| nil         | String ""          | nil | **Y** |
| false       | nil                | nil |   |
| true        | true (TrueClass) | nil | **Y** |
| ""          | String ""          | nil | **Y** |
| "other1"    | String "other"     | nil | **Y** |
| "NaN"       | BigDecimal NaN     | BigDecimal NaN |   |
| "Infinity"  | String "Infinity"  | BigDecimal Infinity  | **Y** |
| "-Infinity" | String "-Infinity" | BigDecimal -Infinity | **Y** |

### Mongoize

| Value to Mongoize | Current        | New | Behavior Changed? |
| ---            | ---            | --- | --- |
| BigDecimal 0   | String "0.0"   | String "0.0"  |   |
| BigDecimal -0  | String "-0.0"  | String "-0.0" |   |
| BigDecimal 1   | String "1"     | String "1"    |   |
| BigDecimal 1.1 | String "1.1"   | String "1.1"  |   |
| nil            | String ""      | nil | **Y** |
| false          | false (FalseClass) | nil | **Y** |
| true           | String "true"  | nil | **Y** |
| ""             | String ""      | nil | **Y** |
| "1"            | String "1"     | String "1"   |   |
| "1.1"          | String "1.1"   | String "1.1" |   |
| "other1"       | String "other" | nil | **Y** |
| String "NaN"         | String "NaN"       | String "NaN"       |   |
| String "Infinity"    | String "Infinity"  | String "Infinity"  |   |
| String "-Infinity"   | String "-Infinity" | String "-Infinity" |   |
| BigDecimal NaN       | String "NaN"       | String "NaN"       |   |
| BigDecimal Infinity  | String "Infinity"  | String "Infinity"  |   |
| BigDecimal -Infinity | String "-Infinity" | String "-Infinity" |   |
| Float::NAN           | String "NaN"       | String "NaN"       |   |
| Float::INFINITY      | String "Infinity"  | String "Infinity"  |   |
| Float::INFINITY * -1 | String "-Infinity" | String "-Infinity" |   |

### CODE REVIEW

**Comment by johnnyshields** [comment]
One more note here, since `numeric?` is only used for BigDecimal, it might make sense to remove it from all the classes where it appears and instead make a private method in BigDecimal (`numeric?(object)`) which evaluates whether the object is numeric. Less monkey patching that way.

**Comment by johnnyshields** [comment]
I've raised MONGOID-4165 to track this.

File: spec/mongoid/extensions/big_decimal_spec.rb [hunk 1]
```diff
@@ -22,18 +70,29 @@
       end
     end
 
-    context "when the value is a float" do
+    context "when the the value is true" do
```
Comment (estolfo) [inline spec/mongoid/extensions/big_decimal_spec.rb]: there's an extra 'the' here

**Comment by estolfo** [comment]
Hi @johnnyshields 
Thanks again for all your work in this PR.
We are ready to merge this into master now. Would you mind rebasing against master so we can double check that the tests pass on travis? Thanks!

**Comment by johnnyshields** [comment]
@estolfo done. Travis CI running now, looks like it should be good.

**Comment by estolfo** [comment]
thanks!

File: spec/mongoid/extensions/object_spec.rb [hunk 2]
```diff
@@ -289,4 +289,15 @@
...
     end
   end
+
+  describe "#numeric?" do
+
+    let(:object) do
+      Object.new
+    end
+
+    it "returns true" do
```
Comment (estolfo) [inline spec/mongoid/extensions/object_spec.rb]: this should say false

File: lib/mongoid/extensions/big_decimal.rb [hunk 3]
```diff
@@ -37,28 +49,25 @@ module ClassMethods
         #
         # @param [ Object ] object The object to demongoize.
         #
-        # @return [ Object ] The object.
+        # @return [ BigDecimal ] A BigDecimal derived from the object or nil.
```
Comment (estolfo) [inline lib/mongoid/extensions/big_decimal.rb]: This should say 
```ruby
[ BigDecimal, nil ]
```

File: lib/mongoid/extensions/big_decimal.rb [hunk 4]
```diff
@@ -37,28 +49,25 @@ module ClassMethods
...
-        # decimal.
+        # Mongoize an object of any type to how it's stored in the db as a String.
         #
         # @example Mongoize the object.
         #   BigDecimal.mongoize(123)
         #
         # @param [ Object ] object The object to Mongoize
         #
-        # @return [ String ] The mongoized object.
+        # @return [ String ] A String representing the object or nil.
```
Comment (estolfo) [inline lib/mongoid/extensions/big_decimal.rb]:
```ruby
[ String, nil ]
```

**Comment by johnnyshields** [comment]
@estolfo I fixed all the notes you raised.

**Comment by estolfo** [comment]
thanks!
